PhysicsTools/NanoAOD/plugins/DisTauTag.cc

```diff
@@ -0,0 +1,366 @@
+    }
+    graphDef_ = tensorflow::loadGraphDef(graphPath_);
+    session_ = tensorflow::createSession(graphDef_);
+  }
+}
+
+void DisTauTag::cleanupTensorFlow() {
+  std::lock_guard<std::mutex> lock(session_mutex_);
+  if (session_) {
+    tensorflow::closeSession(session_);
+    session_ = nullptr;
+  }
+  if (graphDef_) {
+    delete graphDef_;
+    graphDef_ = nullptr;
+  }
+}
+
+template <typename Scalar>
+Scalar DisTauTag::getDeltaPhi(Scalar phi1, Scalar phi2) {
+  static constexpr Scalar pi = boost::math::constants::pi<Scalar>();
+  Scalar dphi = phi1 - phi2;
+  if (dphi > pi)
+    dphi -= 2 * pi;
+  else if (dphi <= -pi)
+    dphi += 2 * pi;
+  return dphi;
+}
+
+void DisTauTag::fill_zero(tensorflow::Tensor& tensor) {
+  size_t size_ = 1;
+  int num_dimensions = tensor.shape().dims();
+  for (int ii_dim = 0; ii_dim < num_dimensions; ii_dim++)
+    size_ = size_ * tensor.shape().dim_size(ii_dim);
+
+  for (size_t ii = 0; ii < size_; ii++)
+    tensor.flat<float>()(ii) = 0.0;
+}
+
+template <typename FeatureT>
+const float DisTauTag::Scale(const Int_t idx, const Float_t value, const bool inner) const {
+  return std::clamp((value - FeatureT::mean.at(idx).at(inner)) / FeatureT::std.at(idx).at(inner),
+                    FeatureT::lim_min.at(idx).at(inner),
+                    FeatureT::lim_max.at(idx).at(inner));
+}
+
+void DisTauTag::saveInputs(const tensorflow::Tensor& tensor,
+                           const std::string& block_name,
+                           const edm::EventID& eventId,
+                           size_t jetIndex) const {
+  const int tau_n = tensor.shape().dim_size(0);
+  const int pf_n = tensor.shape().dim_size(1);
+  const int ftr_n = tensor.shape().dim_size(2);
+
+  std::string json_file_name = "distag_" + std::to_string(eventId.run()) + "_" +
+                               std::to_string(eventId.luminosityBlock()) + "_" + std::to_string(eventId.event()) + "_" +
+                               "jet_" + std::to_string(jetIndex) + ".json";
+
+  std::ofstream json_file(json_file_name.data());
+  if (!json_file.is_open()) {
+    edm::LogError("DisTauTag") << "Failed to open file for saving inputs: " << json_file_name;
+    return;
+  }
+
+  json_file << "{\"" << block_name << "\":[";
+  for (int tau_idx = 0; tau_idx < tau_n; tau_idx++) {
+    json_file << "[";
+    for (int pf_idx = 0; pf_idx < pf_n; pf_idx++) {
+      json_file << "[";
+      for (int ftr_idx = 0; ftr_idx < ftr_n; ftr_idx++) {
+        json_file << std::setprecision(7) << std::fixed << tensor.tensor<float, 3>()(tau_idx, pf_idx, ftr_idx);
+        if (ftr_idx < ftr_n - 1)
+          json_file << ", ";
+      }
+      json_file << "]";
+      if (pf_idx < pf_n - 1)
+        json_file << ", ";
+    }
+    json_file << "]";
+    if (tau_idx < tau_n - 1)
+      json_file << ", ";
+  }
+  json_file << "]}";
+}
+
+void DisTauTag::produce(edm::StreamID, edm::Event& event, const edm::EventSetup& setup) const {
+  auto jets = getHandle(event, jets_token);
```

Comment:
Seems like the `Handle` property is used, so this could be
```suggestion
  auto jets = event.getHandle(jets_token);
```
after which the local `getHandle()` function is not needed anymore.